README.md
# label-distribution-learning-fer-tf
Official Tensorflow Implementation of "Uncertainty-aware Label Distribution Learning for Facial Expression Recognition" paper

# Requirements
python>=3.7

```
pip install tensorflow-gpu==2.4.0
pip install pandas opencv-python scikit-learn matplotlib
pip install git+https://github.com/qubvel/classification_models.git
```

**macos m1**
```shell
# version
numpy=1.22.3
keras=2.11.0
tensorflow-estimator=2.11.0
tensorflow-macos=2.11.0
tensorflow-metal=0.7.0
# pip install
conda install -c apple tensorflow-deps
pip install tensorflow-macos
pip install tensorflow-metal
pip install numpy --upgrade
```

**autodl服务器**

```shell
conda create -n pytf python=3.8.0
conda init bash
conda activate pytf
pip install tensorflow-gpu==2.4.0
pip install pandas opencv-python scikit-learn matplotlib
pip install git+https://github.com/qubvel/classification_models.git
```


# 外接硬盘文件目录描述
```
autodl-nas
    paper1_noise_result
        best_val_10_8563: 噪音标签为10%训练后的模型，已经损失和准确度记录；
        best_val_20_8455: 噪音标签为20%训练后的模型，已经损失和准确度记录；
        best_val_30_8341: 噪音标签为30%训练后的模型，已经损失和准确度记录；
    paper1_origin_result
        best_val_8732: 原始模型训练的结果
        best_val_8755: 原始模型训练的结果
```


# Datasets
Download the [RAF-DB](http://www.whdeng.cn/RAF/model1.html#dataset) dataset and extract the `aligned` folder (contains aligned faces) into `data/rafdb/aligned`
We provide the pseudo valence-arousal as well as the pre-built K-nearest-neighbors for each instance in `train.csv`. The annotation file should have the following columns

| subDirectory_filePath | expression | valence | arousal | knn |
|:---------------------:|:----------:|:-------:|:-------:|:---:|
| ...                   | ...        | ...     | ...     | ... |

The preprocessed data annotations are available at [Data Annotation](https://drive.google.com/drive/folders/1nm91FPPj2Se305GWC-zpxqnNfEQvz2uq?usp=sharing)
    
# Trained model
Coming soon...

# Training 
Download our pretrained backbone on MS-Celeb1M and put the .h5 files into `./pretrained` folder: [Google Drive](https://drive.google.com/drive/folders/1xplUqjQ49ozP5VrG3DbzyZkgZ58y80Wo?usp=sharing)


**Run the following command to train the model on datasets**
```Shell
# Training model with resnet50 backbone
python src/train.py --cfg=config_resnet50_raf --train_data_path=data/rafdb/raf_train.csv --train_image_dir=data/rafdb/aligned 

# Training model with resnet18 backbone
python src/train.py --cfg=config_resnet18_raf --train_data_path=data/rafdb/raf_train.csv --train_image_dir=data/rafdb/aligned 

#Resume training from the latest checkpoint
python train.py --train_data=/path/to/train_data.csv --resume
```
**PyCharm run param**
```Shell
--train_data_path /Volumes/out_disk/datasets/RAF-DB/data/raf_train.csv
--train_image_dir /Volumes/out_disk/datasets/RAF-DB/Image/aligned

--val_data_path 
--val_image_dir 
--cfg
--resume


```



# Evaluation
Run this command to evaluate the trained model on the test set. We use the classification accuracy as our evaluation metric.
```Shell

python src/eval.py --cfg=config_resnet50_raf --trained_weights=trained_weights/trained_resnet50_raf --test_data_path=data/rafdb/raf_test.csv --test_image_dir=data/rafdb/aligned

```

-


# linux Run the model

### network log
{'name': 'Base Config', 'backbone': 'resnet50', 'feature_dim': 512, 'pretrained': 'msceleb', 'input_size': [112, 112], 'pad_size': 4, 'batch_size
': 32, 'num_parallel_calls': -1, 'optimizer': 'adam', 'lr': 0.0001, 'lr_decay': 0.1, 'lr_steps': [10, 30], 'gamma': 0.01, 'num_neighbors': 8, 'la
mb_init': 0.5, 'lamb_lr': 10, 'lamb_beta': 0, 'num_classes': 7, 'class_names': ['Neutral', 'Happy', 'Sad', 'Surprise', 'Fear', 'Disgust', 'Anger'
], 'class_weights': None, 'val_interval': 200, 'save_interval': 2, 'epochs': 60, 'checkpoint_dir': 'weights_checkpoint/resnet50_raf'}

### 服务器上执行代码(aliged)
#### 训练
python src/train.py --cfg=config_resnet50_raf --train_data_path=/home/lz/datasets/RAF-DB/data/raf_train.csv --train_image_dir=/home/lz/datasets/RAF-DB/Image/aligned

python src/train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned

python train.py --cfg=config_resnet50_raf --train_data_path=/root/datasets/RAF-DB/data/raf_train.csv --train_image_dir=/root/datasets/RAF-DB/Image/aligned


#### 验证测试
python src/eval.py --cfg=config_resnet50_raf --trained_weights=/home/lz/code/label-distribution-learning-fer-tf/weights_checkpoint/resnet50_raf/epoch_60  --test_data_path=/home/lz/datasets/RAF-DB/data/raf_test.csv --test_image_dir=/home/lz/datasets/RAF-DB/Image/aligned

### 服务器上执行代码(original)
#### 训练
python src/train.py --cfg=config_resnet50_raf --train_data_path=/home/lz/datasets/RAF-DB/data/raf_train_original.csv --train_image_dir=/home/lz/datasets/RAF-DB/Image/original

恢复训练

python src/train.py --cfg=config_resnet50_raf --train_data_path=/home/lz/datasets/RAF-DB/data/raf_train_original.csv --train_image_dir=/home/lz/datasets/RAF-DB/Image/original --resume

#### 验证测试
python src/eval.py --cfg=config_resnet50_raf --trained_weights=/home/lz/code/label-distribution-learning-fer-tf/aligned_wc/resnet50_raf/epoch_60  --test_data_path=/home/lz/datasets/RAF-DB/data/raf_test_original.csv --test_image_dir=/home/lz/datasets/RAF-DB/Image/original

#### aligned测试orignal的数据
python src/eval.py --cfg=config_resnet50_raf --trained_weights=/home/lz/code/label-distribution-learning-fer-tf/aligned_wc/resnet50_raf/epoch_60  --test_data_path=/home/lz/datasets/RAF-DB/data/raf_test_original.csv --test_image_dir=/home/lz/datasets/RAF-DB/Image/original


## AUTODL服务器

### envs
```shell
matplotlib              3.6.2
numpy                   1.23.5
opencv-python           4.6.0.66
pandas                  1.5.2
tensorflow-estimator    2.4.0
tensorflow-gpu          2.4.0

--cfg=config_resnet50_raf
--train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv 
--train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned

```
# 数据集的处理

## RAF-DB数据集的面部遮挡
**对RAF-DB的测试集图片进行遮挡**
```shell
git clone https://github.com/aqeelanwar/MaskTheFace.git
python mask_the_face.py --path=/root/autodl-tmp/RAF-DB-mask/Image/aligned --mask_type=surgical_blue --color=#ffffff --color_weight=0.5
```
mask_type取值：["surgical", "N95", "KN95", "cloth", "gas", "inpaint", "random", "all"],

color取值：'"#fc1c1a","#177ABC","#94B6D2","#A5AB81","#DD8047","#6b425e","#e26d5a","#c92c48","#6a506d","#ffc900",'
             ' "#ffffff","#000000","#49ff00","#0473e2""'

**此时获取到数据集图像上遮挡，文件的命名也包含遮挡物的类型后缀，所以需要更改csv文件中的文件名**
```shell
cp /root/autodl-tmp/RAF-DB/data/raf_test.csv /root/autodl-tmp/RAF-DB/data/raf_test_gas.csv
// mac命令
sed -i '' "s/.jpg/_gas.jpg/g" /root/autodl-tmp/RAF-DB/data/raf_test_gas.csv
// autodl服务器命令
sed -i "s/.jpg/_gas.jpg/g" /root/autodl-tmp/RAF-DB/data/raf_test_gas.csv

cp /root/autodl-tmp/RAF-DB/data/raf_test.csv /root/autodl-tmp/RAF-DB/data/raf_test_kn95.csv
// mac命令
sed -i '' "s/.jpg/_KN95.jpg/g" /root/autodl-tmp/RAF-DB/data/raf_test_kn95.csv
// autodl服务器命令
sed -i "s/.jpg/_KN95.jpg/g" /root/autodl-tmp/RAF-DB/data/raf_test_kn95.csv


cp /root/autodl-tmp/RAF-DB/data/raf_test.csv /root/autodl-tmp/RAF-DB/data/raf_test_surgical_blue.csv
// mac命令
sed -i '' "s/.jpg/_surgical_blue.jpg/g" /root/autodl-tmp/RAF-DB/data/raf_test_surgical_blue.csv
// autodl服务器命令
sed -i "s/.jpg/_surgical_blue.jpg/g" /root/autodl-tmp/RAF-DB/data/raf_test_surgical_blue.csv
```




## AffectNet数据集的处理
**AffectNet数据集的处理，获取训练接和测试集的csv文件**
```shell
// train
python datasetUtils/affectNet_label.py --image_path=/root/autodl-tmp/AffectNet/train_set/images --label_path=/root/autodl-tmp/AffectNet/train_set/annotations --save_path=/root/autodl-tmp/AffectNet/data/aff_train.csv
// test 
python datasetUtils/affectNet_label.py --image_path=/root/autodl-tmp/AffectNet/val_set/images --label_path=/root/autodl-tmp/AffectNet/val_set/annotations --save_path=/root/autodl-tmp/AffectNet/data/aff_test.csv
```


### RUN the code

**对标论文作者原始代码，使用预训练模型的代码**
```shell
cp /root/autodl-tmp/RAF-DB/pretrained /root/LAL_FER_TF/pretrained
cd /root/LAL_FER_TF
conda activate pytf
git checkout other_code_20230101
python src/train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
python train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
```

**作者论文改造后，使用原始resnet50的代码**
```shell
cd /root/LAL_FER_TF
conda activate pytf
git checkout other_resnet50_train
python train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned --pretrained=imagenet
```

**增加多通道网络后的代码**
```shell
cd /root/LAL_FER_TF
conda activate pytf
git checkout other_resnet50_train
python train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned --pretrained=imagenet --resnetPooling=None

// 增加保存迭代的参数 && 测试数据集
python train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned --pretrained=imagenet --resnetPooling=None --save_interval=2 --val_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv --val_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
// 增加参数验证间隔 【正确的模型训练跑的命令】
python train.py --cfg=config_resnet50_raf --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --train_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned --pretrained=imagenet --resnetPooling=None --save_interval=5 --val_interval=200 --val_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv --val_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned | tee /root/autodl-nas/paper1_origin_result/train_log.txt

```

# 验证
**一次性测试在多个epoch上的准确率**【废弃】
```shell
python eval.py --cfg=config_resnet50_raf --trained_weights=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/epoch_60 --test_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv --test_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned --pretrained=imagenet --resnetPooling=None --trained_weights_dir=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/
```

## 在pycharm中远程连接autodl，验证模型效果的参数配置

**RAF-DB数据集**
```shell
# 正常的验证逻辑
--cfg=config_resnet50_raf
--trained_weights=/root/autodl-nas/paper1_origin_result/best_val_8755
--test_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv
--test_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
--pretrained=imagenet
--resnetPooling=None
--trained_weights_dir=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/

# 10%噪音标签的验证逻辑
--cfg=config_resnet50_raf
--trained_weights=/root/autodl-nas/paper1_noise_result/best_val_10_8563
--test_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv
--test_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
--pretrained=imagenet
--resnetPooling=None
--trained_weights_dir=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/

# 20%噪音标签的验证逻辑
--cfg=config_resnet50_raf
--trained_weights=/root/autodl-nas/paper1_noise_result/best_val_20_8455
--test_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv
--test_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
--pretrained=imagenet
--resnetPooling=None
--trained_weights_dir=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/


# 30%噪音标签的验证逻辑
--cfg=config_resnet50_raf
--trained_weights=/root/autodl-nas/paper1_noise_result/best_val_30_8341
--test_data_path=/root/autodl-tmp/RAF-DB/data/raf_test.csv
--test_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned
--pretrained=imagenet
--resnetPooling=None
--trained_weights_dir=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/


# 面部遮挡的验证逻辑（使用的是正常训练集训练的模型，进行面部遮挡的测试验证）
# 以gas为例
--cfg=config_resnet50_raf
--trained_weights=/root/autodl-nas/paper1_origin_result/best_val_8755
--test_data_path=/root/autodl-tmp/RAF-DB/data/new_raf_test_mask.csv
--test_image_dir=/root/autodl-tmp/RAF-DB/Image/aligned_mask
--pretrained=imagenet
--resnetPooling=None
--trained_weights_dir=/root/autodl-tmp/code/LDL_FER_TF/weights_checkpoint/resnet50_raf/
```



## 噪音标签
**生成不同比例的噪音标签，获取对应的csv文件**
```shell
python datasetUtils/noise_label.py --train_data_path=/root/autodl-tmp/RAF-DB/data/raf_train.csv --dst_data_path=/root/autodl-tmp/RAF-DB/data/ --pre=0.1
```





# 


CHANGELOG.md

#Version


## v0.0.9
**branch**: our_resnet50_train

**change_log**:
1. 增加affectNet数据集处理的脚本

## v0.0.8
**branch**: our_resnet50_train

**change_log**:
1. 增加噪音标签生成脚本

## v0.0.7
**branch**: our_resnet50_train

**change_log**:
1. --save_interval 迭代次数配置


## v0.0.6
**branch**: our_resnet50_train

**change_log**:
1. 每次批次训练，记录损失和准确度
2. **增加多通道网络，构建创新点1的完整代码**


## v0.0.5
**branch**: our_resnet50_train

**change_log**:
1. 增加t-sne简单代码
2、resnet50网络增加参数，决定是否需要全连接层
3、main启动增加参数--resnetPooling决定是需要全连接层

## v0.0.4
**branch**: our_resnet50_train

**change_log**:
1. 根据keras.application构建resnet50


## v0.0.3
**branch**: other_resnet50_train

**change_log**:
1. readme.md增加原始resnet50训练命令
2. train.py兼容原始resnet50训练
3. train.py增加参数pretrained
4. models.py输入维度调整

## v0.0.2_1
**branch**: other_resnet50_train

**change_log**:
1. 增加readme.md

## v0.0.2
**branch**: other_resnet50_train

**change_log**:
1. src中文件移除

## v0.0.1
**branch**: other_resnet50_train

**change_log**:
1. model更改文件位置
2. 增加changeLog.md
3. README.md增加autodl运行命令

# Branch
## other_code_20230101
其他作者的源代码，可以直接执行。


## other_resnet50_train



train.py
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
import tensorflow as tf
import numpy as np
from tensorflow.keras.utils import Progbar
import pandas as pd
from datetime import datetime

from model.models import create_model
from data_utils import *
import utils

import argparse
import sys
sys.path.append("cfg_files")
sys.path.append("model")

gpu_devices = tf.config.experimental.list_physical_devices("GPU")
for device in gpu_devices:
    tf.config.experimental.set_memory_growth(device, True)

def parse_arg(argv=None):
    parser = argparse.ArgumentParser()
    parser.add_argument("--train_data_path", type=str,
                        default="data/rafdb/raf_train.csv",
                        help="Path to the train_data.csv, should have the following columns:\n'subDirectory_filePath,expression,valence,arousal,knn'",
                        required=True)
    parser.add_argument("--train_image_dir", type=str,
                    default="data/rafdb/aligned",
                    help="Path to the directory containing training images",
                    required=True)
    parser.add_argument("--val_data_path", type=str,
                        default=None,
                        help="Path to the validation_data.csv, should have the following columns:\n'subDirectory_filePath,expression'")
    parser.add_argument("--val_image_dir", type=str,
                    default=None,
                    help="Path to the validation_data.csv, should have the following columns:\n'subDirectory_filePath,expression'")
    # parser.add_argument("--pretrained_weights", type=str,
    #                     default=None,
    #                     help="load the pretrained weights of the model in /path/to/model_weights")
    parser.add_argument("--cfg", type=str,
                        default="config_resnet50_raf",
                        help="config file_name")
    parser.add_argument("--pretrained", type=str,
                        default="msceleb",
                        help="if msceleb, use pretrained model; Or None, use keras.application.resnet50")
    parser.add_argument("--resnetPooling", type=str,
                        default='avg',
                        help="if avg, max, None")
    parser.add_argument("--resume",
                        action= "store_true",
                        help="Resume training from the last checkpoint")

    parser.add_argument("--save_interval", type=int,
                        default=0,
                        help="Resume training from the last checkpoint")

    parser.add_argument("--val_interval", type=int,
                        default=0,
                        help="Resume training from the last checkpoint")

    parser.add_argument("--batch_size", type=int,
                        default=32,
                        help="Resume training from the last checkpoint")



    args = parser.parse_args(argv)
    return args

def train_step(model, optimizer, x_batch_train, y_batch_train, x_batch_aux, config,
               global_labels, lamb_hat, lamb_optim, dcm_loss=None,
               loss_weights_dict=None, class_weights_dict=None,
               knn_weights=None,
               lamb=None, idx=None, neighbor_idx=None):
    if loss_weights_dict is None:
        loss_weights_dict = {
            'emotion_loss': 1,
            'va_loss': 1,
        }

    sample_weights = utils.create_sample_weights(y_batch_train[0], class_weights_dict)
    # Get neighbor prediction

    B, K, H, W, _ = x_batch_aux.shape
    x_batch_aux = tf.reshape(x_batch_aux, shape=(-1, H, W, 3))
    feat_aux, preds_aux = model(x_batch_aux, training=True)

    preds_aux = tf.reshape(preds_aux, shape=(B, K, -1))  # shape (B,K,C)
    feat_aux = tf.reshape(feat_aux, shape=(B, K, -1))  # shape (B,K,C)

    with tf.GradientTape() as tape:
        feat, preds = model(x_batch_train, training=True)
        attention_weights = model.weighting_net((feat), (feat_aux), training=True)

        # construct label distribution
        emotion_cls_pred = preds
        emotion_cls_true = y_batch_train[0]


        if idx is not None:
            lamb = tf.gather(lamb_hat, idx)
            lamb = tf.sigmoid(lamb)

        emotion_cls_true = utils.construct_target_distribution(emotion_cls_true, (preds_aux), knn_weights,
                                                               (attention_weights), lamb=lamb)

        emotion_loss = utils.CELoss(emotion_cls_true, emotion_cls_pred, sample_weights)

        dcm_loss_value = dcm_loss(feat, y_batch_train[0], feat_aux, tf.gather(global_labels, neighbor_idx),
                                    lamb_hat=lamb_hat,
                                    indices = tf.concat([idx, tf.reshape(neighbor_idx, -1)], axis=0))

        total_loss = emotion_loss + config.gamma * dcm_loss_value

        optimizing_variables = model.trainable_variables + [lamb_hat]
        gradients = tape.gradient(total_loss, optimizing_variables)
        gradients, lamb_hat_grad = gradients[:-1], gradients[-1]

        optimizer.apply_gradients(
            [(grad, var) for (grad, var) in zip(gradients, optimizing_variables) if grad is not None])

        lamb_optim.apply_gradients([(lamb_hat_grad, lamb_hat)])

        dcm_loss.update_centers(dcm_loss.compute_grads())

    return preds, total_loss, {'emotion_loss': emotion_loss,
                               'lamb': tf.reduce_mean(tf.reduce_mean(tf.sigmoid(tf.gather(lamb_hat, idx))))}


def test_step(model, x_batch, y_batch):
    feat, preds = model(x_batch, training=False)

    emotion_cls_pred = preds
    emotion_cls_true = y_batch[0]
    emotion_loss = tf.keras.losses.SparseCategoricalCrossentropy()(emotion_cls_true, emotion_cls_pred)

    total_loss = emotion_loss

    return preds, total_loss, {'emotion_loss': emotion_loss}






def train(model, optimizer, train_dataset, global_labels, config,
          val_dataset=None, epochs=5, load_checkpoint=False,
          loss_weights_dict=None,
          class_weights=None):
    # define metrics for controlling process
    train_loss = tf.keras.metrics.Mean('train_loss', dtype=tf.float32)
    val_loss = tf.keras.metrics.Mean('val_loss', dtype=tf.float32)
    train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy('train_accuracy')
    val_accuracy = tf.keras.metrics.SparseCategoricalAccuracy('val_accuracy')

    batches_per_epoch = tf.data.experimental.cardinality(train_dataset).numpy()
    ckpt_dir = config.checkpoint_dir

    # init values
    dcm_loss = utils.DiscriminativeLoss(config.num_classes, feature_dim=config.feature_dim)
    lamb_hat = tf.Variable(tf.zeros(shape=(len(global_labels), 1), dtype=tf.float32) + config.lamb_init)
    lamb_optim = tf.keras.optimizers.SGD(config.lamb_lr, config.lamb_beta)
    class_weights_dict = None
    if class_weights:
        if type(class_weights) == bool:
            class_weights = sklearn.utils.class_weight.compute_class_weight("balanced",
                                                                            classes=np.unique(global_labels.numpy()),
                                                                            y=global_labels.numpy())
            print(class_weights)
        class_weights_dict = {i:v for i,v in enumerate(class_weights)}
    best_val = 0
    iter_count = 0
    val_interval = config.val_interval
    save_interval = config.save_interval



    # setup checkpoint manager
    checkpoint = tf.train.Checkpoint(step=tf.Variable(0), model=model, optimizer=optimizer,
                                     dcm_loss=dcm_loss)
    manager = tf.train.CheckpointManager(
        checkpoint, directory="./checkpoints", max_to_keep=1
    )
    if load_checkpoint:
        status = checkpoint.restore(manager.latest_checkpoint)
        if manager.latest_checkpoint:
            print("Restored from {}".format(manager.latest_checkpoint))
        else:
            print("Initializing from sratch....")
    else:
        print("Initializing from scratch...")

    iter_count = checkpoint.step.numpy()

    # todo: 记录迭代次数下，损失和准确度
    df = pd.DataFrame(columns=['time', 'epoch', 'loss', 'accuracy', 'lamb', 'val_loss', 'val_accuracy'])  # 列名
    df.to_csv(f'{ckpt_dir}/train_log.csv', index=False)  # 路径可以根据需要更改

    for epoch in range(epochs):
        train_loss.reset_states()
        train_accuracy.reset_states()

        print("Epoch {}".format(int(iter_count / batches_per_epoch) + 1))
        print("LR: ", float(optimizer.learning_rate(optimizer.iterations)))

        # Training
        pb_i = Progbar(batches_per_epoch, width=30, interval=0.5,
                       stateful_metrics=['total_loss', 'emotion_loss', 'emotion_acc', 'avg_lamb'])
        save_lamb = 0
        for x_batch_train, va_regression_true, emotion_cls_true, x_batch_aux, knn_weights, idx, neighbor_idx in train_dataset:
            checkpoint.step.assign_add(1)
            iter_count += 1
            curr_epoch = int(iter_count / batches_per_epoch)

            lamb = None #we use trainable lambda
            y_batch_train = (emotion_cls_true, va_regression_true)
            preds, total_loss, loss_dict = train_step(model, optimizer, x_batch_train, y_batch_train, x_batch_aux, config,
                                                      global_labels, lamb_hat, lamb_optim, dcm_loss=dcm_loss,
                                                      class_weights_dict=class_weights_dict,
                                                      knn_weights=knn_weights,
                                                      lamb=lamb, idx=idx, neighbor_idx=neighbor_idx,
                                                      )
            train_loss(total_loss)
            emotion_acc = train_accuracy(emotion_cls_true, preds)

            pb_i.add(1, [('total_loss', total_loss.numpy()),
                         ('emotion_accuracy', emotion_acc),
                         ('avg_lamb', loss_dict['lamb'])])

            save_lamb = loss_dict['lamb']

            if iter_count % val_interval == 0 and val_dataset is not None:
                val_loss.reset_states()
                val_accuracy.reset_states()
                for x_batch, va_regression_true, emotion_cls_true in val_dataset:
                    y_batch = (emotion_cls_true, va_regression_true)
                    preds, total_loss, loss_dict = test_step(model, x_batch, y_batch)
                    val_loss(total_loss)  # update metric
                    val_accuracy(emotion_cls_true, preds)  # update metric
                acc = val_accuracy.result()
                print("\n---Iterations: {}, Val loss: {:.4}, Val Acc: {:.4}".format(iter_count,val_loss.result(), acc))
                if (acc > best_val):
                    model.save_weights(f"{ckpt_dir}/best_val/Model")
                    print("====Best validation model saved!====")
                    best_val = acc

        save_path = manager.save()
        if (curr_epoch) % save_interval == 0:
            model.save_weights(f'{ckpt_dir}/epoch_' + str(curr_epoch) + '/Model')

        print('End of Epoch: {}, Iter: {}, Train Loss: {:.4}, Emotion Acc: {:.4}'.format(curr_epoch, iter_count,
                                                                                         train_loss.result(),
                                                                                         train_accuracy.result()))


        # Validation
        if val_dataset is not None:
            val_loss.reset_states()
            val_accuracy.reset_states()

            for x_batch, va_regression_true, emotion_cls_true in val_dataset:
                y_batch = (emotion_cls_true, va_regression_true)
                preds, total_loss, loss_dict = test_step(model, x_batch, y_batch)

                val_loss(total_loss)  # update metric
                val_accuracy(emotion_cls_true, preds)  # update metric

            print('Val loss: {:.4},  Val accuracy: {:.4}'.format(val_loss.result(), val_accuracy.result()))
            print('===================================================')

            if (val_accuracy.result() > best_val):
                model.save_weights(f"{ckpt_dir}/best_val/Model")
                print("====Best validation model saved!====")
                best_val = val_accuracy.result()
        print()

        # todo: 记录每次迭代的损失和准确度
        list = ["%s" % datetime.now(), epoch,
                '%.4f' % train_loss.result(),
                '%.4f' % train_accuracy.result(),
                '%.4f' % save_lamb,
                '%.4f' % val_loss.result(),
                '%.4f' % val_accuracy.result(),
                ]
        data = pd.DataFrame([list])
        data.to_csv(f'{ckpt_dir}/train_log.csv', mode='a', header=False, index=False)  # mode设为a,就可以向csv文件追加数据了

    return model


def main(train_data_path, image_dir, config, val_data_path=None, val_image_dir = None):
    assert (val_image_dir is not None) or (val_data_path is None)

    model = create_model(config)
    print("Model created!")

    train_dataset = get_train_dataset(train_data_path, image_dir, config)
    optimizer = utils.get_optimizer(train_dataset, config)
    val_dataset = get_test_dataset(val_data_path, val_image_dir, config) if val_data_path is not None else None
    print("Dataset loaded!")

    train_data = pd.read_csv(train_data_path)
    global_labels = tf.constant(train_data['expression'], dtype=tf.int32)


    print("Start training...")
    train(model, optimizer, train_dataset, global_labels, config,
          val_dataset=val_dataset,
          epochs=config.epochs,
          load_checkpoint=args.resume,
          class_weights=config.class_weights)

if __name__ == '__main__':
    args = parse_arg()


    config = __import__(args.cfg).config
    if args.pretrained != config.pretrained:
        config.pretrained = args.pretrained
        config.feature_dim = 2048
    if args.resnetPooling == 'None':
        config.resnetPooling = None
    else:
        config.resnetPooling = args.resnetPooling

    if args.save_interval != 0:
        print('Config_save_interval: {}'.format(args.save_interval))
        config.save_interval = args.save_interval

    if args.val_interval != 0:
        print('Config_val_interval: {}'.format(args.val_interval))
        config.val_interval = args.val_interval

    if args.batch_size != 32:
        print('batch_size: {}'.format(args.batch_size))
        config.batch_size = args.batch_size

    print(config.__dict__)

    main(train_data_path= args.train_data_path, image_dir=args.train_image_dir, config= config,
         val_data_path=args.val_data_path, val_image_dir=args.val_image_dir)


